compiler/src/main/kotlin/visitors/typechecker.kt

```diff
@@ -54,16 +54,31 @@ class TypeChecker(symbolTable: Table) : ScopedASTVisitor(symbolTable = symbolTab
         node.setType(BooleanType)
     }
 
-    override fun visit(node: ArrayLookupExpr) {
+    override fun visit(node: LookupExpr) {
         super.visit(node)
 
-        val arrayType = node.arr.getType()
+        val type = node.target.getType()
 
-        if (arrayType is ArrayType) {
-            node.setType(arrayType.subtype)
-        } else {
-            ErrorLogger += TypeError(node.ctx, "Cannot lookup in expression of type $arrayType")
-            node.setType(UndeterminedType)
+        when (type) {
+            is ArrayType -> {
+                node.lookupType = LookupExprType.ARRAY
+                node.setType(type.subtype)
+            }
+
+            LocalNeighbourhoodType -> {
```

Comment:
Should this case and `StateType` be prefixed with `is`?

Reply:
No. `is ArrayType` is neccessary because ArrayType is a class, `LocalNeighbourhoodType` and `StateType` are objects, so we can just test if the type is one of those specific objects.